pkg/webreg/graphviz.go

```diff
@@ -2,183 +2,224 @@ package webreg
+			"attrs": func(e edge) string {
+				var a []string
+				if e.srcType == subgraphType {
+					a = append(a, fmt.Sprintf("ltail=cluster_%d", e.srcGraph))
+				}
+				if e.dstType == subgraphType {
+					a = append(a, fmt.Sprintf("lhead=cluster_%d", e.dstGraph))
+					a = append(a, "minlen=2")
+				}
+				if len(a) == 0 {
+					return ""
+				}
+				return fmt.Sprintf("[%s]", strings.Join(a, " "))
+			},
+		}).Parse(graphvizTemplateStr))
+	graphvizSGTemplate = template.Must(
+		template.New("webreg_sg").Funcs(template.FuncMap{
+			"label": func(g graph, i int) string {
+				return g.subgraphs[i].label
+			},
+			"nodes": func(g graph, i int) []int {
+				return g.subgraphs[i].nodes
+			},
+		}).Parse(graphvizSGTemplateStr))
+)
+
+// addSubgraph creates a sub-graph for a sequence of steps
+func (b *graphBuilder) addSubgraph(label string, steps []api.TestStep) int {
+	sg := subgraph{label: label}
+	hasSrc := b.edge.srcType != noEdge
+	edge := len(b.graph.edges)
+	b.edge.dstType = subgraphType
+	if len(steps) == 0 {
+		b.addNode(&sg, node{label: "Intentionally left blank"})
 	}
-	lastIndex := -1
-	lastType := subgraphType
-	for _, step := range root {
-		var currNode *node
-		currSG := -1
+	for _, step := range steps {
 		if step.LiteralTestStep != nil {
-			currNode = &node{label: step.As, linkable: false}
+			b.addNode(&sg, node{label: step.As})
 		} else if step.Reference != nil {
-			currNode = &node{label: *step.Reference, linkable: true}
+			b.addNode(&sg, node{label: *step.Reference, linkable: true})
 		} else if step.Chain != nil {
-			mainGraph, currSG = addSubgraph(mainGraph, *step.Chain, chains[*step.Chain].Steps, chains, !isRoot, true, false)
-		}
-		// create new edge
-		var newIndex int
-		var newType edgeType
-		if currNode != nil {
-			newType = nodeType
-			mainGraph.nodes = append(mainGraph.nodes, *currNode)
-			newIndex = len(mainGraph.nodes) - 1
-			if sg.firstNode == -1 {
-				sg.firstNode = newIndex
-			}
-			sg.nodes = append(sg.nodes, newIndex)
-		} else if currSG != -1 {
-			newType = subgraphType
-			newIndex = currSG
-			if sg.firstNode == -1 {
-				sg.firstNode = mainGraph.subgraphs[newIndex].firstNode
-			}
-			sg.subgraphs = append(sg.subgraphs, newIndex)
-		}
-		if lastIndex != -1 {
-			newEdge := edge{
-				dst:     newIndex,
-				dstType: newType,
-			}
-			newEdge.srcType = lastType
-			newEdge.src = lastIndex
-			mainGraph.edges = append(mainGraph.edges, newEdge)
+			i := b.addSubgraph(*step.Chain, b.chains[*step.Chain].Steps)
+			sg.subgraphs = append(sg.subgraphs, i)
 		}
-		lastIndex = newIndex
-		lastType = newType
 	}
-	// this is used to identify a node that can be linked from is this sg is a src for an edge
-	if lastType == nodeType {
-		sg.lastNode = lastIndex
-	} else {
-		sg.lastNode = mainGraph.subgraphs[lastIndex].lastNode
+	i := len(b.graph.subgraphs)
+	b.graph.subgraphs = append(b.graph.subgraphs, sg)
+	b.edge.srcType = subgraphType
+	b.edge.srcGraph = i
+	if hasSrc && edge != len(b.graph.edges) {
```

Comment:
Please add a comment here explaining what's the reason of being of this `if`

Reply:
You caught me, this is the most subtle section of the code.  Comments added.